clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
BTW, your usecase also suggests that the change can be qualified as an API change. The ticket for this change would look like "Allow KafkaConsumer to avoid network IO when there is  data available to return".